pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -1309,3 +1337,89 @@ func initializeAnnotations(cd *hivev1.ClusterDeployment) {
 		cd.Annotations = map[string]string{}
 	}
 }
+
+// mergePullSecrets takes pull secretes in to account and returns the merged pull request
+func (r *ReconcileClusterDeployment) mergePullSecrets(cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) (string, error) {
+
+	// For code readability let's call the pull secret in cluster deployment config as local pull secret
+	localPullSecret, _ := controllerutils.LoadSecretData(r.Client, cd.Spec.PullSecret.Name, cd.Namespace, corev1.DockerConfigJsonKey)
+
+	// Check if global pull secret from env as it comes from hive config
+	globalPullSecret := os.Getenv("GLOBAL_PULL_SECRET")
+
+	switch {
+	case globalPullSecret != "" && localPullSecret != "":
+		// Merge local pullSecret and globalPullSecret. If both pull secrets have same registry name
+		// then the merged pull secret will have registry secret from local pull secret
+		pullSecret, err := controllerutils.MergeJsons(globalPullSecret, localPullSecret, cdLog)
+		if err != nil {
+			errMsg := "unable to merge global pull secret with local pull secret"
+			cdLog.WithError(err).Error(errMsg)
+			return "", fmt.Errorf("%s : %v", errMsg, err)
+		}
+		return pullSecret, nil
+	case globalPullSecret != "":
+		return globalPullSecret, nil
+	case localPullSecret != "":
+		return localPullSecret, nil
+	default:
+		errMsg := "clusterdeployment must specify pull secret since hiveconfig does not specify a global pull secret"
+		cdLog.Error(errMsg)
+		return "", fmt.Errorf("%s", errMsg)
+	}
+}
+
+// updatePullSecretInfo adds pull secret information in cluster deployment and cluster deployment status
+func (r *ReconcileClusterDeployment) updatePullSecretInfo(pullSecret string, cd *hivev1.ClusterDeployment, cdLog log.FieldLogger) error {
+	var secretName string
+	var noExistingPullSecretObj bool
+
+	existingPullSecretObj := &corev1.Secret{}
+
+	if cd.Status.PullSecret.Name != "" {
+		secretName = cd.Status.PullSecret.Name
+		err := r.Get(context.TODO(), types.NamespacedName{Name: secretName, Namespace: cd.Namespace}, existingPullSecretObj)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				noExistingPullSecretObj = true
+			} else {
+				cdLog.WithError(err).Error("Error getting pull secret from cluster deployment")
+				return err
+			}
+		}
+	} else {
+		secretName = apihelpers.GetResourceName(cd.Name, "pull-secret")
```

Comment:
`GetResourceName(cd.Name, "pull-secret")`  does not add random strings to the name, hence it is crashing with existing pull secret name created in `generatePullSecret() (*corev1.Secret, error) {` 
`"error creating pull secret object: secrets \"lalatendum-tectonic-4-pull-secret\" already exists"` 

Hence I need to use `GetResourceName(cd.Name, "pull-secret"+utilrand.String(8))` cc @dgoodwin

Reply:
merged-pull-secret perhaps?

Reply:
@dgoodwin calling it `merged-pull-secret` wont be correct as it because when global pull secret is not available, it wont be not be not merged exactly.

Reply:
It is still merged. In that case, it is merged with an empty pull secret.

Reply:
used `merged` in the updated code